File: examples/java/basic/Forwarder.java
/*
 * $Id$
 *
 * Copyright 2006,2007,2010 The SimGrid Team
 * All rights reserved. 
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. 
 */

import simgrid.msg.*;

public class Forwarder extends simgrid.msg.Process {
    
   public void main(String[] args) throws JniException, NativeException {
      if (args.length < 3) {	 
	 Msg.info("Forwarder needs 3 arguments (input mailbox, first output mailbox, last one)");
	 Msg.info("Got "+args.length+" instead");
	 System.exit(1);
      }
      int input = Integer.valueOf(args[0]).intValue();		
      int firstOutput = Integer.valueOf(args[1]).intValue();		
      int lastOutput = Integer.valueOf(args[2]).intValue();		
      
      int taskCount = 0;
      int slavesCount = lastOutput - firstOutput + 1;
      Msg.info("Receiving on 'slave_"+input+"'");
      while(true) {
	 Task t = Task.receive("slave_"+input);	
	 
	 if (t instanceof FinalizeTask) {
	    Msg.info("Got a finalize task. Let's forward that we're done.");
	    
	    for (int cpt = firstOutput; cpt<=lastOutput; cpt++) {
	       Task tf = new FinalizeTask();
	       tf.send("slave_"+cpt);
	    }
	    break;
	 }
	 BasicTask task = (BasicTask)t;
	 int dest = firstOutput + (taskCount % slavesCount);
	 
	 Msg.info("Sending \"" + task.getName() + "\" to \"slave_" + dest + "\"");
	 task.send("slave_"+dest);
	    
	 taskCount++;
      }
      
	 
      Msg.info("I'm done. See you!");
   }
}

File: examples/scala/masterslave/Forwarder.scala
/*
 * Copyright (c) 2006-2013. The SimGrid Team.
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. 
 */

package masterslave

import Stream._
import org.simgrid.msg.Host
import org.simgrid.msg.Msg
import org.simgrid.msg.MsgException
import org.simgrid.msg.Task
import org.simgrid.msg.Process

class Forwarder(host:Host, name:String, args:Array[String]) extends Process(host,name,args) {

   def main(args: Array[String]){
      if (args.length < 3) {	 
	 Msg.info("Forwarder needs 3 arguments (input mailbox, first output mailbox, last one)")
	 Msg.info("Got "+args.length+" instead")
	 System.exit(1)
      }
      val input = args(0).toInt		
      val firstOutput = args(1).toInt		
      val lastOutput = args(2).toInt		
      
      var taskCount = 0
      val slavesCount = lastOutput - firstOutput + 1
      Msg.info("Receiving on 'slave_"+input+"'")
      var cont = true
      
      continually({Task.receive("slave_"+input)})
        .takeWhile(!_.isInstanceOf[FinalizeTask])
        .foreach(task => {
          val dest = firstOutput + (taskCount % slavesCount)
          Msg.info("Sending \"" + task.getName() + "\" to \"slave_" + dest + "\"")
          task.send("slave_"+dest)
          taskCount += 1
        })

      Msg.info("Got a finalize task. Let's forward that we're done.")
      for (cpt <- firstOutput to lastOutput) {
        val tf = new FinalizeTask()
        tf.send("slave_"+cpt)
      }
	 
      Msg.info("I'm done. See you!")
   }
}

